### PR TITLE
[mlir] introduce transform.collect_matching

### PR DESCRIPTION
**Pull request by ftynse**
Introduce a new match combinator into the transform dialect. This
operation collects all operations that are yielded by a satisfactory
match into its results. This is a simpler version of `foreach_match`
that can be inserted directly into existing transform scripts.

### CODE REVIEW

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 1]
```diff
@@ -822,6 +823,137 @@ matchBlock(Block &block, Operation *op, transform::TransformState &state,
...
+//===----------------------------------------------------------------------===//
+
+DiagnosedSilenceableFailure
+transform::CollectMatchingOp::apply(transform::TransformRewriter &rewriter,
+                                    transform::TransformResults &results,
+                                    transform::TransformState &state) {
+  auto matcher = SymbolTable::lookupNearestSymbolFrom<FunctionOpInterface>(
+      getOperation(), getMatcher());
+  if (matcher.isExternal()) {
+    return emitDefiniteFailure()
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Transform/IR/TransformOps.cpp:853]: Could this be verified in the op verifier?

File: mlir/include/mlir/Dialect/Transform/IR/TransformOps.td [hunk 2]
```diff
@@ -459,6 +459,36 @@ def NumAssociationsOp : TransformDialectOp<"num_associations",
   }];
 }
 
+def CollectMatchingOp : TransformDialectOp<"collect_matching", [
+    DeclareOpInterfaceMethods<MemoryEffectsOpInterface>,
+    DeclareOpInterfaceMethods<SymbolUserOpInterface>,
+    DeclareOpInterfaceMethods<TransformOpInterface>]> {
+  let summary = "Collects all payload ops that match the given named matcher";
+  let description = [{
+    Collects operations nested under `root` or other payload IR objects that
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:468]: I would mention that `root` itself is not taken into account when matching ops.

File: mlir/include/mlir/Dialect/Transform/IR/TransformOps.td [hunk 3]
```diff
@@ -459,6 +459,36 @@ def NumAssociationsOp : TransformDialectOp<"num_associations",
...
+    match the given matcher expressed as a named sequence. The matcher sequence
+    must accept exactly one argument that it is not allowed to modify. It must
+    yield as many values as this op has results. Each of the yielded values must
+    be associated with exactly one payload object. If any operation in the
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:472]: What's the reason for requiring that exactly 1 op is associated with each result handle?

File: mlir/include/mlir/Dialect/Transform/IR/TransformOps.td [hunk 4]
```diff
@@ -459,6 +459,36 @@ def NumAssociationsOp : TransformDialectOp<"num_associations",
...
+    matcher sequence produces a silenceable failure, the matcher advances to the
+    next payload operation in the walk order without finishing the sequence.
+
+    The results of this operation are constructed by concatenating values
```
Comment (matthias-springer) [inline mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:476]: nit: I would rephrase this along the lines of:

`The i-th result of this operation is constructed by concatenating the i-th yielded payload IR objects of all successful matcher sequence applications. All results are guaranteed to be mapped to the same number of payload IR objects.`

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 5]
```diff
@@ -822,6 +823,137 @@ matchBlock(Block &block, Operation *op, transform::TransformState &state,
...
+      symbolTable.lookupNearestSymbolFrom(getOperation(), getMatcher()));
+  if (!matcherSymbol ||
+      !isa<TransformOpInterface>(matcherSymbol.getOperation()))
+    return emitError() << "unresolved matcher symbol " << getMatcher();
+
+  ArrayRef<Type> argumentTypes = matcherSymbol.getArgumentTypes();
+  if (argumentTypes.size() != 1 ||
+      !isa<TransformHandleTypeInterface>(argumentTypes[0])) {
+    return emitError()
+           << "expected the matcher to take one operation handle argument";
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Transform/IR/TransformOps.cpp:925]: value handles are also allowed?

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 6]
```diff
@@ -822,6 +823,137 @@ matchBlock(Block &block, Operation *op, transform::TransformState &state,
...
+  }
+  if (!matcherSymbol.getArgAttr(
+          0, transform::TransformDialect::kArgReadOnlyAttrName)) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Transform/IR/TransformOps.cpp:928]: Is there an easy way to check that matcher does not modify the payload?

File: mlir/test/Dialect/Transform/test-interpreter.mlir [hunk 7]
```diff
@@ -2167,3 +2167,32 @@ transform.sequence failures(propagate) {
   transform.yield 
 }
 }
+
+// -----
```
Comment (matthias-springer) [inline mlir/test/Dialect/Transform/test-interpreter.mlir:2171]: successful test case is missing

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 8]
```diff
@@ -822,6 +823,137 @@ matchBlock(Block &block, Operation *op, transform::TransformState &state,
...
+        rawResults[i].push_back(mapping[0]);
+      }
+      return WalkResult::advance();
+    });
+    if (walkResult.wasInterrupted())
+      return std::move(*maybeFailure);
+    assert(!maybeFailure && "failure set but the walk was not interrupted");
+
+    for (auto &&[opResult, rawResult] :
+         llvm::zip(getOperation()->getResults(), rawResults)) {
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Transform/IR/TransformOps.cpp:899]: nit: `zip_equal`

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 9]
```diff
@@ -822,6 +823,137 @@ matchBlock(Block &block, Operation *op, transform::TransformState &state,
...
+        DEBUG_MATCHER(DBGS_MATCHER() << "matcher " << matcher.getName()
+                                     << " failed: " << diag.getMessage());
+        return WalkResult::advance();
+      }
+
+      // If succeeded, collect results.
+      for (auto &&[i, mapping] : llvm::enumerate(mappings)) {
+        if (mapping.size() != 1) {
+          maybeFailure.emplace(emitSilenceableError()
+                               << "result #" << i << ", associated with "
```
Comment (matthias-springer) [inline mlir/lib/Dialect/Transform/IR/TransformOps.cpp:885]: This sounds like incorrect usage of the transform op. Why is this not a definite error?

File: mlir/include/mlir/Dialect/Transform/IR/TransformOps.td (hunk 2 above)
Comment (ftynse) [reply mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:468]: It is taken into account, the implementaiton does a plain `walk`. Do we want an option that doesn't include root?

File: mlir/include/mlir/Dialect/Transform/IR/TransformOps.td (hunk 3 above)
Comment (ftynse) [reply mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:472]: So we the user can expect a certain number of associations in the result and how they map to associations of the operand. For example, for future co-iteration. We could eventually relax that to all "calls" having equal number of associations for each result, but probably not more than that.

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp (hunk 1 above)
Comment (ftynse) [reply mlir/lib/Dialect/Transform/IR/TransformOps.cpp:853]: No, at verification time we allow external symbols. They may be "resolved" later by merging modules or by providing a library module to the interpreter.

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp (hunk 9 above)
Comment (ftynse) [reply mlir/lib/Dialect/Transform/IR/TransformOps.cpp:885]: For me, definite errors are a graceful equivalent of `abort()`. They mean that the payload IR was modified in an irrecoverable way and we should immediately stop the interpreter, and the parent pass. This is fully recoverable. The surrounding op (usually sequence) may choose to suppress the error or report it.

File: mlir/include/mlir/Dialect/Transform/IR/TransformOps.td (hunk 2 above)
Comment (nicolasvasilache) [reply mlir/include/mlir/Dialect/Transform/IR/TransformOps.td:469]: `-> Collects operations nested under `root` (included)` ?

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp [hunk 10]
```diff
@@ -822,6 +823,137 @@ matchBlock(Block &block, Operation *op, transform::TransformState &state,
...
+        DBGS_MATCHER() << "matching ";
+        op->print(llvm::dbgs(),
+                  OpPrintingFlags().assumeVerified().skipRegions());
+        llvm::dbgs() << " @" << op << "\n";
+      });
+
+      // Try matching.
+      SmallVector<SmallVector<MappedValue>> mappings;
+      DiagnosedSilenceableFailure diag =
+          matchBlock(matcher.getFunctionBody().front(), op, state, mappings);
```
Comment (nicolasvasilache) [inline mlir/lib/Dialect/Transform/IR/TransformOps.cpp:872]: FMI: what is the use case for matchers returning definite failures?
Can they modify IR and leave it in an undefined state?

**Review by nicolasvasilache — APPROVED** [review]
Nice !

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp (hunk 10 above)
Comment (ftynse) [reply mlir/lib/Dialect/Transform/IR/TransformOps.cpp:872]: Normally, _match_ operations should not. But we don't internally differentiate match ops from other transform ops internally right now.

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

File: mlir/lib/Dialect/Transform/IR/TransformOps.cpp (hunk 6 above)
Comment (ftynse) [reply mlir/lib/Dialect/Transform/IR/TransformOps.cpp:928]: The presence of this attribute triggers the verifier on named sequence to check that ops nested in the named sequence don't have a write effect on the payload IR. The effect specification may be wrong, but there is little we can do in that case, other than cloning the payload before and comparing that indeed nothing has changed under an -enable-even-more-expensive checks flag.

**Review by matthias-springer — APPROVED** [review]
